### PR TITLE
Add mish activation function

### PR DESCRIPTION
**Pull request by Adnios**
See issus: https://github.com/pytorch/pytorch/issues/58375

### CODE REVIEW

**Comment by ezyang** [comment]
i simple man, i see structured, i likey (letting @jbschlosser do the actual review)

**Review by jbschlosser — COMMENTED** [review]
Wow, thank you for the beautiful and thorough work! Couple small things and this will be good to go :)

* Please add tests for the C++ module / functional forms to `test/cpp/api/modules.cpp` and `test/cpp/api/functional.cpp`. Once built, they can be run with `build/bin/test_api`.
* **Optionally**, adding an `OpInfo` entry for `mish()` to the `op_db` list in `torch/testing/_internal/common_methods_invocations.py` will add a bunch more generic testing.
* I don't think it's worth doing for this PR, but our `softplus` implementation accepts a `beta` parameter; in the future, it might be worth exposing this as configurable for `mish` as well.

File: docs/source/scripts/build_activation_images.py [hunk 1]
```diff
@@ -37,6 +37,7 @@
     'RReLU',
     'SELU',
     'SiLU',
+    'MISH',
```
Comment (jbschlosser) [inline docs/source/scripts/build_activation_images.py:40]: `MISH` -> `Mish` should get the docs building

**Comment by Adnios** [comment]
> Wow, thank you for the beautiful and thorough work! Couple small things and this will be good to go :)
> 
> * Please add tests for the C++ module / functional forms to `test/cpp/api/modules.cpp` and `test/cpp/api/functional.cpp`. Once built, they can be run with `build/bin/test_api`.
> * **Optionally**, adding an `OpInfo` entry for `mish()` to the `op_db` list in `torch/testing/_internal/common_methods_invocations.py` will add a bunch more generic testing.
> * I don't think it's worth doing for this PR, but our `softplus` implementation accepts a `beta` parameter; in the future, it might be worth exposing this as configurable for `mish` as well.

Thanks for your advice. I will implement some of these as soon as possible.

**Review by jbschlosser — COMMENTED** [review]
Looks great! One tiny lint-related issue and we're good to go :)

File: aten/src/ATen/native/Activation.h [hunk 2]
```diff
@@ -54,6 +54,8 @@ DECLARE_DISPATCH(activation_fn, glu_stub);
 DECLARE_DISPATCH(activation_backward_fn, glu_backward_stub);
 DECLARE_DISPATCH(structured_activation_fn, silu_stub);
 DECLARE_DISPATCH(activation_backward_fn, silu_backward_stub);
+DECLARE_DISPATCH(structured_activation_fn , mish_stub);
```
[the remark this answers is not in the document]
Comment (Adnios) [reply aten/src/ATen/native/Activation.h:57]: Thanks! What should I do? Let's keep it the same or change the code.

[the remark this answers is not in the document]
Comment (digantamisra98) [reply aten/src/ATen/native/Activation.h:57]: simply remove the space after `structured_activation_fn` before the comma. :)

[the remark this answers is not in the document]
Comment (Adnios) [reply aten/src/ATen/native/Activation.h:57]: haha, Thanks @digantamisra98 . It is my mistake.

**Comment by JackCaoG** [comment]
@Adnios Could you mark the test with `onlyOnCPUAndCUDA` so it won't be run with XLA? Xla usually won't try to follow the same runtime error message throw by pytorch core so we asked developer not to run this kind of these against pt/xla.

**Comment by Adnios** [comment]
> @Adnios Could you mark the test with `onlyOnCPUAndCUDA` so it won't be run with XLA? Xla usually won't try to follow the same runtime error message throw by pytorch core so we asked developer not to run this kind of these against pt/xla.

Hello @JackCaoG, thanks for your help!

I just check the log message of `pytorch_xla_linux_bionic_py3_6_clang9_test` and find the `test_silu_inplace_overlap_xla` and `test_softplus_inplace_overlap_xla` are skiped. The `test_mish_inplace_overlap_xla` should also be skiped.  I think I will try to add `@onlyOnCPUAndCUDA`.
```
May 22 19:39:42   test_silu_inplace_overlap_xla (__main__.TestNNDeviceTypeXLA) ... skip (0.002s)
May 22 19:39:42   test_smooth_l1_loss_vs_huber_loss_xla (__main__.TestNNDeviceTypeXLA) ... skip (0.003s)
May 22 19:39:42   test_softmax_64bit_indexing_xla (__main__.TestNNDeviceTypeXLA) ... skip (0.002s)
May 22 19:39:42   test_softmax_bfloat16_xla (__main__.TestNNDeviceTypeXLA) ... skip (0.001s)
May 22 19:43:27   test_softmax_results_xla_float32 (__main__.TestNNDeviceTypeXLA) ... ok (224.887s)
May 22 19:43:27   test_softmax_xla_float16 (__main__.TestNNDeviceTypeXLA) ... skip (0.005s)
May 22 19:43:27   test_softmax_xla_float32 (__main__.TestNNDeviceTypeXLA) ... skip (0.003s)
May 22 19:43:27   test_softplus_inplace_overlap_xla (__main__.TestNNDeviceTypeXLA) ... skip (0.001s)
```

File: aten/src/ATen/native/cpu/Activation.cpp [hunk 3]
```diff
@@ -632,6 +632,45 @@ void silu_backward_kernel(TensorIterator& iter) {
...
+            iter,
+            [](scalar_t x) {
+                return x * std::tanh(std::log1p(std::exp(x)));
+            },
+            [kOneVec](Vec256<scalar_t> x_vec) {
+              return x_vec * x_vec.exp().log1p().tanh();
+            });
+      });
+}
+
```
Comment (venkateshtata) [inline aten/src/ATen/native/cpu/Activation.cpp:648]: Hi @Adnios, I am still understanding the PyTorch code architecture and came across your commit which is really interesting. Could you please explain me why do we have a mish_kernel and a mish_backward_kernel ? and why did you define math_mish_backward. Thank you.

Comment (Adnios) [reply aten/src/ATen/native/cpu/Activation.cpp:648]: Hello, @venkateshtata. Sorry for the delay.
- `mish_kernel` and `mish_backward_kernel` are the C++ implementations of `mish` which are used by `torch._C._nn.mish` in `torch/nn/functional.py`
- `math_mish_backward` may be not necessary for is commit. But there may be a potential problem. See: https://github.com/pytorch/pytorch/pull/49439 https://github.com/pytorch/xla/issues/2652

I refer to a lot of codes of `softplus` and `silu` when I implement `mish`.
The blog of ezyang is very helpful to understand the PyTorch. http://blog.ezyang.com/2019/05/pytorch-internals/
Hope this helps.

Comment (venkateshtata) [reply aten/src/ATen/native/cpu/Activation.cpp:648]: Thank you

File: torch/nn/modules/activation.py [hunk 4]
```diff
@@ -398,6 +398,41 @@ def extra_repr(self) -> str:
...
 
+class Mish(Module):
+    r"""Applies the Mish function, element-wise.
+    Mish: A Self Regularized Non-Monotonic Neural Activation Function.
+
+    .. math::
+        \text{Mish}(x) = x * \text{Tanh}(\text{Softplus}(x))
+
+    .. note::
+        See `Mish: A Self Regularized Non-Monotonic Neural Activation Function <https://arxiv.org/abs/1908.08681>`.
```
Comment (jbschlosser) [inline torch/nn/modules/activation.py:409]: Looks like the paper link didn't render correctly 
(see [rendered version](https://13676694-65600975-gh.circle-artifacts.com/0/docs/generated/torch.nn.Mish.html)).

I think you just need an underscore at the end:
```
See `Mish: A Self Regularized Non-Monotonic Neural Activation Function <https://arxiv.org/abs/1908.08681>`_
```

Comment (Adnios) [reply torch/nn/modules/activation.py:409]: Thanks. I will fix it in the next commit. I will commit again after the checks finished in order to make sure that there is no other mistakes.

**Review by jbschlosser — APPROVED** [review]

File: aten/src/ATen/native/Activation.cpp [hunk 5]
```diff
@@ -306,6 +320,23 @@ Tensor math_silu_backward(
...
+  mish_backward_stub(iter.device_type(), iter);
+  return grad_input;
+}
+
+Tensor math_mish_backward(
+    const Tensor& grad_output,
+    const Tensor& input) {
+  auto input_tanh_softplus = at::tanh(at::softplus(input));
+  auto input_sigmoid = at::sigmoid(input);
+  return grad_output * (input_tanh_softplus + (input * input_sigmoid * (1 - input_tanh_softplus * input_tanh_softplus)));
```
Comment (vadimkantorov) [inline aten/src/ATen/native/Activation.cpp:337]: will this create many intermediate copies?

**Pull request closed without merging**